packages/vite/src/node/plugins/workerImportMetaUrl.ts

```diff
@@ -0,0 +1,127 @@
+import { ResolvedConfig } from '../config'
+import { Plugin } from '../plugin'
+import { getAssetHash, fileToUrl } from './asset'
+import {
+  cleanUrl,
+  injectQuery,
+  multilineCommentsRE,
+  singlelineCommentsRE
+} from '../utils'
+import { TransformPluginContext } from 'rollup'
+import path from 'path'
+import { bundleWorkerScript } from './worker'
+import { ENV_PUBLIC_PATH } from '../constants'
+import MagicString from 'magic-string'
+import { parse as parseUrl, URLSearchParams } from 'url'
+
+const WorkerFileId = 'worker_url_file'
+
+interface URLWithImportMetaUrl {
+  start: number
+  end: number
+  workerConstruct: string
+  file: string
+}
+
+function parseWorkerRequest(id: string): Record<string, string> | null {
+  const { search } = parseUrl(id)
+  if (!search) {
+    return null
+  }
+  return Object.fromEntries(new URLSearchParams(search.slice(1)))
+}
+
+function workerImportMetaUrl(
+  ctx: TransformPluginContext,
+  code: string,
+  id: string,
+  options?: { ssr?: boolean }
+): URLWithImportMetaUrl[] {
+  const result: URLWithImportMetaUrl[] = []
+  if (
+    code.includes('new Worker') &&
+    code.includes('new URL') &&
+    code.includes(`import.meta.url`)
+  ) {
+    const importMetaUrlRE =
+      /\bnew\s+[window\.|self\.|globalThis\.]*(Worker|SharedWorker)\s*\(\s*(new\s+[window\.|self\.|globalThis\.]*URL\s*\(\s*('[^']+'|"[^"]+"|`[^`]+`)\s*,\s*import\.meta\.url\s*\))/g
```

Comment:
[What this mean?](https://regexper.com/#%2F%5Cbnew%5Cs%2B%5Bwindow%5C.%7Cself%5C.%7CglobalThis%5C.%5D*%28Worker%7CSharedWorker%29%5Cs*%5C%28%5Cs*%28new%5Cs%2B%5Bwindow%5C.%7Cself%5C.%7CglobalThis%5C.%5D*URL%5Cs*%5C%28%5Cs*%28'%5B%5E'%5D%2B'%7C%22%5B%5E%22%5D%2B%22%7C%60%5B%5E%60%5D%2B%60%29%5Cs*%2C%5Cs*import%5C.meta%5C.url%5Cs*%5C%29%29%2Fg)
<img width="1419" alt="image" src="https://user-images.githubusercontent.com/7699524/146499453-8de23a5d-890c-4795-881b-e73e25958aa2.png">